core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceConcurrencyTest.java

```diff
@@ -0,0 +1,351 @@
+    ListeningScheduledExecutorService executor = newTestExecutor();
+    TestDataSupplier supplier = new TestDataSupplier(true, executor);
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "a:b:c",
+            supplier,
+            AuthType.PASSWORD,
+            new TestCredentialFactory(),
+            executor,
+            keyPairFuture,
+            newRateLimiter());
+    assertThat(supplier.counter.get()).isEqualTo(0);
+
+    // Attempt to retrieve data, ensure we wait for success
+    ListenableFuture<List<Object>> allData =
+        Futures.allAsList(
+            executor.submit(instance::getSslData),
+            executor.submit(instance::getSslData),
+            executor.submit(instance::getSslData));
+
+    List<Object> d = allData.get();
+    assertThat(d.get(0)).isNotNull();
+    assertThat(d.get(1)).isNotNull();
+    assertThat(d.get(2)).isNotNull();
+    assertThat(supplier.counter.get()).isEqualTo(1);
+
+    for (int i = 0; i < 10; i++) {
+      // Call forceRefresh simultaneously
+      ListenableFuture<List<Object>> all =
+          Futures.allAsList(
+              executor.submit(instance::forceRefresh),
+              executor.submit(instance::forceRefresh),
+              executor.submit(instance::forceRefresh));
+      try {
+        all.get();
+      } catch (Exception e) {
+      }
+
+      ListenableFuture<List<Object>> allData2 =
+          Futures.allAsList(
+              executor.submit(instance::getSslData),
+              executor.submit(instance::getSslData),
+              executor.submit(instance::getSslData));
+      try {
+        allData2.get();
+      } catch (Exception e) {
+      }
+      assertThat(supplier.counter.get()).isEqualTo(2 + i);
+      Thread.sleep(100);
+    }
+  }
+
+  @Test
+  public void testForceRefreshDeadlock() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    ListenableFuture<KeyPair> keyPairFuture =
+        Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    ListeningScheduledExecutorService executor = newTestExecutor();
+    TestDataSupplier supplier = new TestDataSupplier(true, executor);
+    List<CloudSqlInstance> instances = new ArrayList<>();
+
+    final int instanceCount = 5;
+
+    for (int i = 0; i < instanceCount; i++) {
+      instances.add(
+          new CloudSqlInstance(
+              "a:b:instance" + i,
+              supplier,
+              AuthType.PASSWORD,
+              new TestCredentialFactory(),
+              executor,
+              keyPairFuture,
+              newRateLimiter()));
+    }
+
+    assertThat(supplier.counter.get()).isEqualTo(0);
+
+    // Get SSL Data for each instance, forcing the first refresh
+    instances.forEach(i -> i.getSslData());
+
+    assertThat(supplier.counter.get()).isEqualTo(instanceCount);
+
+    List<Thread> threads =
+        instances.stream()
+            .map(
+                (inst) -> {
```

Comment:
Could we pull this lambda out into a method to make its purpose clear?

Reply:
Done.